File: worker/src/features/room/services/NovaStt.ts
import type { Bindings } from "../../../core/types";
import type { SttCallbacks } from "../types";

type NovaMessage = {
  type: string;
  channel?: { alternatives: [{ transcript: string }] };
  is_final?: boolean;
  speech_final?: boolean;
};

const NOVA_PARAMS: Record<string, string> = {
  encoding: "linear16",
  sample_rate: "16000",
  channels: "1",
  interim_results: "true",
  punctuate: "true",
  smart_format: "true",
  endpointing: "300",
  utterance_end_ms: "1000",
};

export class NovaStt {
  private ws: WebSocket | null = null;
  private pending = "";

  constructor(
    private env: Bindings,
    private roomId: string,
    private callbacks: SttCallbacks,
  ) {}

  async connect(sourceLang: string): Promise<void> {
    const { url, headers } = this.buildConnection(sourceLang);

    const resp = await fetch(url, { headers: { Upgrade: "websocket", ...headers } });
    if (resp.status !== 101) {
      const body = await resp.text().catch(() => "");
      throw new Error(`Nova-3 WS ${resp.status}: ${body}`);
    }

    this.ws = resp.webSocket!;
    this.ws.accept();
    this.ws.addEventListener("message", (e) => this.handleMessage(e.data));
    this.ws.addEventListener("error", (e) => console.error(`[room:${this.roomId}] Nova error:`, e));
  }

  sendAudio(data: ArrayBuffer) {
    if (this.ws?.readyState === WebSocket.OPEN) this.ws.send(data);
  }

  close() {
    this.ws?.close();
    this.ws = null;
  }

  // --- internal ---

  private buildConnection(sourceLang: string): { url: string; headers: Record<string, string> } {
    if (this.env.DEEPGRAM_API_KEY) return this.buildDirectDeepgram(sourceLang);
    return this.buildCfGateway();
  }

  private buildDirectDeepgram(sourceLang: string): { url: string; headers: Record<string, string> } {
    const url = new URL("https://api.deepgram.com/v1/listen");
    url.searchParams.set("model", "nova-3");
    url.searchParams.set("language", sourceLang);
    this.applyParams(url);
    return { url: url.toString(), headers: { Authorization: `Token ${this.env.DEEPGRAM_API_KEY}` } };
  }

  private buildCfGateway(): { url: string; headers: Record<string, string> } {
    const url = new URL(
      `https://gateway.ai.cloudflare.com/v1/${this.env.CF_ACCOUNT_ID}/${this.env.CF_AI_GATEWAY_ID}/workers-ai`,
    );
    url.searchParams.set("model", "@cf/deepgram/nova-3");
    url.searchParams.set("language", "en");
    this.applyParams(url);
    return { url: url.toString(), headers: { "cf-aig-authorization": `Bearer ${this.env.CF_API_TOKEN}` } };
  }

  private applyParams(url: URL) {
    for (const [key, value] of Object.entries(NOVA_PARAMS)) url.searchParams.set(key, value);
  }

  private async handleMessage(data: string | ArrayBuffer) {
    if (typeof data !== "string") return;

    const msg = this.parse(data);
    if (!msg) return;

    const transcript = msg.channel?.alternatives?.[0]?.transcript?.trim() ?? "";

    if (!msg.is_final && !msg.speech_final) {
      if (!transcript) return;
      this.pending = transcript;
      this.callbacks.onInterim(transcript);
    } else {
      const finalTranscript = transcript || this.pending;
      this.pending = "";
      if (finalTranscript) this.callbacks.onFinal(finalTranscript, Date.now());
    }
  }

  private parse(data: string): NovaMessage | null {
    try {
      const msg = JSON.parse(data) as NovaMessage;
      return msg.type === "Results" ? msg : null;
    } catch {
      return null;
    }
  }
}

File: worker-rs/src/room/nova_stt.rs
use worker::*;

const NOVA_PARAMS: &[(&str, &str)] = &[
    ("encoding", "linear16"),
    ("sample_rate", "16000"),
    ("channels", "1"),
    ("interim_results", "true"),
    ("punctuate", "true"),
    ("smart_format", "true"),
    ("endpointing", "300"),
    ("utterance_end_ms", "1000"),
];

pub struct NovaStt {
    pub ws: WebSocket,
}

impl NovaStt {
    pub async fn connect(env: &Env, _room_id: &str, source_lang: &str) -> Result<Self> {
        let (url, auth_headers) = Self::build_connection(env, source_lang)?;

        let mut req = Request::new(&url, Method::Get)?;
        let headers = req.headers_mut()?;
        headers.set("Upgrade", "websocket")?;
        for (key, value) in &auth_headers {
            headers.set(key, value)?;
        }

        let mut resp = Fetch::Request(req).send().await?;

        if resp.status_code() != 101 {
            let body = resp.text().await.unwrap_or_default();
            return Err(Error::RustError(format!("Nova-3 WS {}: {}", resp.status_code(), body)));
        }

        let ws = resp.websocket().ok_or_else(|| Error::RustError("No WebSocket in response".into()))?;
        ws.accept()?;

        Ok(Self { ws })
    }

    pub fn send_audio(&self, data: &[u8]) {
        let _ = self.ws.send_with_bytes(data);
    }

    pub fn close(&self) {
        let _ = self.ws.close::<&str>(None, None);
    }

    // --- internals ---

    fn build_connection(env: &Env, source_lang: &str) -> Result<(String, Vec<(String, String)>)> {
        if let Ok(api_key) = env.secret("DEEPGRAM_API_KEY").map(|s| s.to_string()) {
            return Self::build_direct_deepgram(source_lang, &api_key);
        }
        Self::build_cf_gateway(env)
    }

    fn build_direct_deepgram(source_lang: &str, api_key: &str) -> Result<(String, Vec<(String, String)>)> {
        let mut url = Url::parse("https://api.deepgram.com/v1/listen")
            .map_err(|e| Error::RustError(e.to_string()))?;

        url.query_pairs_mut().append_pair("model", "nova-3");
        url.query_pairs_mut().append_pair("language", source_lang);
        Self::apply_params(&mut url);

        let headers = vec![("Authorization".into(), format!("Token {}", api_key))];
        Ok((url.to_string(), headers))
    }

    fn build_cf_gateway(env: &Env) -> Result<(String, Vec<(String, String)>)> {
        let account_id = env.var("CF_ACCOUNT_ID")?.to_string();
        let gateway_id = env.var("CF_AI_GATEWAY_ID")?.to_string();
        let api_token = env.secret("CF_API_TOKEN")?.to_string();

        let base = format!(
            "https://gateway.ai.cloudflare.com/v1/{}/{}/workers-ai",
            account_id, gateway_id
        );
        let mut url = Url::parse(&base).map_err(|e| Error::RustError(e.to_string()))?;

        url.query_pairs_mut().append_pair("model", "@cf/deepgram/nova-3");
        url.query_pairs_mut().append_pair("language", "en");
        Self::apply_params(&mut url);

        let headers = vec![("cf-aig-authorization".into(), format!("Bearer {}", api_token))];
        Ok((url.to_string(), headers))
    }

    fn apply_params(url: &mut Url) {
        let mut pairs = url.query_pairs_mut();
        for (key, value) in NOVA_PARAMS {
            pairs.append_pair(key, value);
        }
    }
}
